flink-ml-lib/src/main/java/org/apache/flink/ml/clustering/kmeans/KMeans.java

```diff
@@ -307,24 +266,41 @@ public void processElement2(StreamRecord<DenseVector[]> streamRecord) throws Exc
 
         @Override
         public void onEpochWatermarkIncremented(
-                int epochWatermark, Context context, Collector<Tuple2<Integer, DenseVector>> out)
+                int epochWatermark,
+                Context context,
+                Collector<Tuple2<Integer[], DenseVector[]>> out)
                 throws Exception {
             DenseVector[] centroidValues =
                     Objects.requireNonNull(
                             OperatorStateUtils.getUniqueElement(centroids, "centroids")
                                     .orElse(null));
+
+            DenseVector[] newCentroids = new DenseVector[centroidValues.length];
+            int[] counts = new int[centroidValues.length];
+            for (int i = 0; i < centroidValues.length; i++) {
+                newCentroids[i] = new DenseVector(centroidValues[i].size());
+            }
+
             for (DenseVector point : points.get()) {
                 int closestCentroidId =
                         findClosestCentroidId(centroidValues, point, distanceMeasure);
-                output.collect(new StreamRecord<>(Tuple2.of(closestCentroidId, point)));
+
+                BLAS.axpy(1.0, point, newCentroids[closestCentroidId]);
+                counts[closestCentroidId]++;
             }
 
+            output.collect(
+                    new StreamRecord<>(
+                            Tuple2.of(
+                                    Arrays.stream(counts).boxed().toArray(Integer[]::new),
```

Comment:
`Arrays.stream(...)` is generally much slower than an iterative approach as of now [1]. If the iterative approach is not too complex to write, how about we use the iterative approach here?

[1] https://stackoverflow.com/questions/27925954/is-arrays-streamarray-name-sum-slower-than-iterative-approach